File: CommandLine_TicTacToe.cpp
#include <iostream>
#include <numeric>
#include <limits>
#include <array>

void displayBoard(const std::array<uint8_t, 9>& board) {
	char t[] { '-', 'X', 'O' };
	std::cout << std::endl <<
		t[board[0]] << " " << t[board[1]] << " " << t[board[2]] << std::endl <<
		t[board[3]] << " " << t[board[4]] << " " << t[board[5]] << std::endl <<
		t[board[6]] << " " << t[board[7]] << " " << t[board[8]] << std::endl;
}

int main(void) {
	std::cout <<
		"Welcome to Tic Tac Toe!\n"
		"Enter the number of the place you want to play!\n"
		"1 2 3\n"
		"4 5 6\n"
		"7 8 9" << std::endl;

	// 0 = empty, 1 = X, 2 = O
	std::array<uint8_t, 9> board {};

	// false = X, true = O
	// uint8_t(player) + 1 converts bool (X=0, O=1) to same format as board array (X=1, O=2)
	bool player = false;

	// Main loop
	while (true) {
		displayBoard(board);

		// Retrieve input
		bool retryInput;
		do {
			retryInput = false;

			std::cout << "Player " << (player ? 'O' : 'X') << ": ";
			int input;
			std::cin >> input;
			std::cin.ignore(std::numeric_limits<int>::max(), '\n');

			// Validate input
			input--;
			if (input < 0 || input >= 9) {
				std::cout << "Please enter a number between 1 and 9!" << std::endl;
				retryInput = true;
			}
			else if (board[input]) {
				std::cout << "That space is taken! Choose another space!" << std::endl;
				retryInput = true;
			}
			else board[input] = uint8_t(player) + 1;
		} while (retryInput);


		// Check for winning board state
		bool win = false;
		for (int i = 0; i < 3; i++) {
			win |= ( // Check row i
				uint8_t(player) + 1 == board[i * 3 + 0] &&
				uint8_t(player) + 1 == board[i * 3 + 1] &&
				uint8_t(player) + 1 == board[i * 3 + 2]
				) || ( // Check column i
				uint8_t(player) + 1 == board[0 + i] &&
				uint8_t(player) + 1 == board[3 + i] &&
				uint8_t(player) + 1 == board[6 + i]
			);
		}
		// Check for both diagonals
		win |= (
			uint8_t(player) + 1 == board[0] &&
			uint8_t(player) + 1 == board[4] &&
			uint8_t(player) + 1 == board[8]
			) || (
			uint8_t(player) + 1 == board[2] &&
			uint8_t(player) + 1 == board[4] &&
			uint8_t(player) + 1 == board[6]
		);

		// Display winning text
		if (win) {
			displayBoard(board);
			std::cout << "Player " << (player ? 'O' : 'X') << " Wins! Press Enter to Escape!";
			while (std::cin.get() != '\n');
			break;
		}


		// Display draw text
		if (std::accumulate(board.begin(), board.end(), 0) == 13) {
			displayBoard(board);
			std::cout << "Draw! Press Enter to Escape!";
			while (std::cin.get() != '\n');
			break;
		}


		// Change player before looping
		player = !player;
	}

	
	return 0;
}

File: CommandLine_TicTacToe.py
def display(board = [0, 0, 0, 0, 0, 0, 0, 0, 0]):
    t = ['-', 'X', 'O']
    print(
'''
{} {} {}
{} {} {}
{} {} {}'''.format(t[board[0]], t[board[1]], t[board[2]],\
                   t[board[3]], t[board[4]], t[board[5]],\
                   t[board[6]], t[board[7]], t[board[8]]))
    return

#main program
print(
'''
Welcome to Tic Tac Toe!
Enter the number of the place you want to play!
1 2 3
4 5 6
7 8 9''')

#0 = empty, 1 = X, 2 = O
Board = [0, 0, 0, 0, 0, 0, 0, 0, 0]

#False = X, True = O
player = False

#main loop
while(True):
    display(Board)

    takeInput = True
    while(takeInput):
        takeInput = False

        try:
            Input = int(input("Player " + ("O" if player else "X") + ": ")) - 1
        except:
            print("Please enter a number between 1 and 9!")
            takeInput = True
            continue

        #validate input
        if(Input < 0 or Input >= 9):
            print("Please enter a number between 1 and 9!")
            takeInput = True
        elif(Board[Input]):
            print("That space is taken! Choose another space1")
            takeInput = True
        else:
            Board[Input] = 2 if player else 1

    #check for winning board state
    win = False
    for i in range(0, 3):
        currentPlayer = 2 if player else 1
        win = win or ((\
            currentPlayer == Board[i * 3 + 0] and\
            currentPlayer == Board[i * 3 + 1] and\
            currentPlayer == Board[i * 3 + 2]\
        ) or (\
            currentPlayer == Board[i + 0] and\
            currentPlayer == Board[i + 3] and\
            currentPlayer == Board[i + 6]))
        
    win = win or ((\
        currentPlayer == Board[0] and\
        currentPlayer == Board[4] and\
        currentPlayer == Board[8]\
    ) or (\
        currentPlayer == Board[2] and\
        currentPlayer == Board[4] and\
        currentPlayer == Board[6]))

    #display winning text
    if win:
        display(Board)
        input("Player " + ("O" if player else "X") + " wins! Press Enter to escape!")
        quit()
    #display draw text
    elif sum(Board) == 13:
        input("Draw! Press enter to escape!")
        quit()

    #swap player before looping
    player = not player
